### PR TITLE
Infinite loop on Java 1.8 compilation when use of lot of Generics

### PR DESCRIPTION
**Pull request by trancexpress**
Adjust `ReturnStatement.resolve()` to only call `TypeBinding.isProperType()` in the presence of a lambda. This avoids unnecessary heavy computations when there are many generic arguments.

Fixes: #177

## Author checklist

- [ ] I have thoroughly tested my changes
- [ ] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [ ] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by trancexpress** [comment]
@stephan-herrmann since you added the code, any idea if the cached value should be invalidated at any point?

**Comment by stephan-herrmann** [comment]
That stack trace in #177 looks weird. Like an infinite recursion, which is exactly what the flag `inRecursiveFunction` should prevent. If that fails, I wouldn't see how caching could possibly help.
I'll have to debug the reproducer to say more. Have you tried creating a JUnit from it?

**Comment by trancexpress** [comment]
> That stack trace in #177 looks weird. Like an infinite recursion, which is exactly what the flag `inRecursiveFunction` should prevent.
Its more complicated than that, as there are different `TypeVariableBinding` objects on the stack trace (the code doesn't actually recurse in the same object and so the flag doesn't help). I'm not convinced its an endless loop, maybe just n<sup>2</sup> (or worse) complexity?

> Have you tried creating a JUnit from it?
No. The reproduction project is quite big, I've not tried to reduce it yet. Its hard to follow what the code does in this project as well.

**Comment by stephan-herrmann** [comment]
> Its more complicated than that, as there are different `TypeVariableBinding` objects on the stack trace (the code doesn't actually recurse in the same object and so the flag doesn't help). I'm not convinced its an endless loop, maybe just n2 (or worse) complexity?

Well, to begin with the class being compiled has a level of complexity that I haven't seen before: `AccordCardBase` has 45 type parameters each bounded by a parameterized type referencing up-to 45 type variables a piece.

Generally, `isProperType()` performs very little computation, so caching can only possibly become relevant in case the same types are checked thousands of times over and over again. I wonder if any compilation step will be able to perform well for such program.

As for a JUnit, I'm sure it only needs the headers of class `AccordCardBase` plus referenced types, and this invocation:
```
public CARDPAYMENTMEANS getCardPaymentMeans() {
  return cardPaymentMeans;
}
```
`ReturnStatement.resolveType()` needs to check `isProperType()` and off we go!

Use type `AccordCardBase` in a lambda requiring type inference, and you might get close to solving the halting problem.

**Comment by stephan-herrmann** [comment]
And here's a change that trivially fixes this exact problem:
```
diff --git a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReturnStatement.java b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReturnStatement.java
index 077c566..2e380b5 100644
--- a/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReturnStatement.java
+++ b/org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/ReturnStatement.java
@@ -346,7 +346,7 @@
        if (methodType == null)
                return;
 
-       if (methodType.isProperType(true) && lambda != null) {
+       if (lambda != null && methodType.isProperType(true)) {
                // ensure that type conversions don't leak a preliminary local type:
                if (lambda.updateLocalTypes())
                        methodType = lambda.expectedResultType();
```
As the question `isProperType()` is only relevant in type inference situation, it can be avoided completely in programs that contain nothing requiring type inference (like, e.g., a lambda).

Hopefully, people using such monster types are wise enough to diligently and explicitly specify all type arguments everywhere, as anything triggering type inference will be a land mine.

**Comment by trancexpress** [comment]
> As for a JUnit, I'm sure it only needs the headers of class `AccordCardBase` plus referenced types, and this invocation:

That is still a lot of types for a JUnit test. You are sure you want us to add one? I can make an archive that is used during the test, of course.

> And here's a change that trivially fixes this exact problem:

Thank you, that works! I'll adjust the PR once we have clarified whether we want to add a big reproducer as a test.

File: org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/LambdaExpressionsTest.java [hunk 1]
```diff
@@ -7399,6 +7399,51 @@ public void testBug529197_003() {
...
+/**
+ * Test that using many generic type arguments doesn't result in a compiler hang.
+ * See: https://github.com/eclipse-jdt/eclipse.jdt.core/issues/177
+ */
+public void testManyGenericsHangGh177() {
+	this.runConformTest(
+			new String[] {
+					"C0.java",
+					"""
+					public class C0
```
[the remark this answers is not in the document]
Comment (stephan-herrmann) [reply org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/compiler/regression/LambdaExpressionsTest.java:7411]: Cool, yes this should come very close to the original.

I suggest to move it to XLargeTest2.

Ideally we'd add a performance-related assertion, but in the past such assertions caused some trouble, so I'm fine with adding it as a regular regression test, just in case and as a help for debugging should we see it hang again.

**Comment by trancexpress** [comment]
Thanks for reviewing Stephan!
